**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: JUNIOR ISRAEL UBRI GONZALEZ
*   MATRÍCULA: A01841481


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [1]:
pip install pymysql

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 2.3 MB/s eta 0:00:00


`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [2]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [3]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE runtime > 300
""")

pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series
3,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
4,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
query = sqla.text("""
    SELECT
        100.0 * SUM(available_globally = 0) / COUNT(available_globally) AS porcentaje_solo_EU
    FROM movie
    WHERE available_globally IS NOT NULL
""")

pd.read_sql(query, conn)

,porcentaje_solo_EU
0,83.57175


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [6]:
query = sqla.text("""
    SELECT DISTINCT locale
    FROM movie
    WHERE locale IS NOT NULL
""")

pd.read_sql(query, conn)

,locale
0,en


In [7]:
query = sqla.text("""
    SELECT COUNT(*) AS cantidad_locale_null
    FROM movie
    WHERE locale IS NULL
""")

pd.read_sql(query, conn)

,cantidad_locale_null
0,11255


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [8]:
query = sqla.text("""
    SELECT COUNT(*) AS titulos_diferentes
    FROM movie
    WHERE title <> original_title
""")

pd.read_sql(query, conn)

,titulos_diferentes
0,3947


**Comparación de resultados:**

La cantidad de registros con locale = NULL es de 11,255, mientras que la cantidad de películas cuyos campos title y original_title son diferentes es de 3,947. Por lo tanto, los resultados no coinciden.

In [9]:
query = sqla.text("""
    SELECT *
    FROM movie
    WHERE locale IS NULL
      AND (
          title = original_title
          OR title IS NULL
          OR original_title IS NULL
      )
""")

pd.read_sql(query, conn)

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,1,2024-01-01,2024-01-01,b'\x01',None,None,2024-03-08,110,Damsel
1,2,2024-01-01,2024-01-01,b'\x01',None,None,2024-01-12,107,Lift
2,5,2024-01-01,2024-01-01,b'\x00',None,None,None,92,The Super Mario Bros. Movie
3,10,2024-01-01,2024-01-01,b'\x00',None,None,None,98,The Boss Baby
4,16,2024-01-01,2024-01-01,b'\x01',None,None,2023-11-21,107,Leo
...,...,...,...,...,...,...,...,...,...
7303,11776,2024-01-01,2024-01-01,None,None,None,None,94,Faithfully Yours
7304,11779,2024-01-01,2024-01-01,None,None,None,None,85,King of Clones
7305,11780,2024-01-01,2024-01-01,None,None,None,None,112,Black Clover: Sword of the Wizard King
7306,11781,2024-01-01,2024-01-01,None,None,None,None,100,You Do You


Los resultados no coinciden directamente. Se identificaron 11,255 registros con locale igual a NULL, mientras que 3,947 películas presentan títulos diferentes entre title y original_title. Al analizar los registros restantes, se obtuvieron 7,308 casos, caracterizados principalmente por presentar también un valor NULL en original_title, aunque sí disponen de información en el campo title. La suma de ambos grupos (3,947 + 7,308) corresponde a los 11,255 registros con locale = NULL.

Por lo tanto, no es válida la suposición de que un valor NULL en locale indique necesariamente que la película está en un idioma diferente del inglés. En una parte importante de estos registros no existe información en original_title que permita realizar la comparación. En consecuencia, los valores NULL de locale deben interpretarse como información de idioma/región no registrada o desconocida, y no como evidencia suficiente de que la película pertenece a otro idioma.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [15]:
query = sqla.text("""
    SELECT
        movie.title,
        view_summary.cumulative_weeks_in_top10
    FROM view_summary
    JOIN movie
        ON view_summary.movie_id = movie.id
    WHERE view_summary.movie_id IS NOT NULL
    ORDER BY view_summary.cumulative_weeks_in_top10 DESC
    LIMIT 1
""")

pd.read_sql(query, conn)

,title,cumulative_weeks_in_top10
0,The Boss Baby,22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [16]:
query = sqla.text("""
    SELECT
        tv_show.title,
        COUNT(season.id) AS cantidad_temporadas
    FROM tv_show
    JOIN season
        ON tv_show.id = season.tv_show_id
    GROUP BY tv_show.id, tv_show.title
    ORDER BY cantidad_temporadas DESC
    LIMIT 5
""")

pd.read_sql(query, conn)

,title,cantidad_temporadas
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [17]:
query = sqla.text("""
    SELECT DISTINCT
        start_date,
        end_date
    FROM view_summary
    WHERE duration = 'SEMI_ANNUALLY'
    ORDER BY start_date
""")

pd.read_sql(query, conn)

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [18]:
query = sqla.text("""
    SELECT
        tv_show.title,
        season.season_number,
        season.title AS season_title,
        view_summary.views
    FROM tv_show
    JOIN season
        ON tv_show.id = season.tv_show_id
    JOIN view_summary
        ON season.id = view_summary.season_id
    WHERE tv_show.title = 'Grey''s Anatomy'
      AND view_summary.duration = 'SEMI_ANNUALLY'
      AND view_summary.start_date = '2024-01-01'
      AND view_summary.end_date = '2024-06-30'
    ORDER BY view_summary.views DESC
""")

pd.read_sql(query, conn)

,title,season_number,season_title,views
0,Grey's Anatomy,1,Grey's Anatomy: Season 1,3600000
1,Grey's Anatomy,2,Grey's Anatomy: Season 2,3100000
2,Grey's Anatomy,3,Grey's Anatomy: Season 3,2900000
3,Grey's Anatomy,5,Grey's Anatomy: Season 5,2900000
4,Grey's Anatomy,4,Grey's Anatomy: Season 4,2900000
5,Grey's Anatomy,6,Grey's Anatomy: Season 6,2800000
6,Grey's Anatomy,7,Grey's Anatomy: Season 7,2700000
7,Grey's Anatomy,8,Grey's Anatomy: Season 8,2600000
8,Grey's Anatomy,9,Grey's Anatomy: Season 9,2500000
9,Grey's Anatomy,10,Grey's Anatomy: Season 10,2400000


**Interpretación de los resultados:**

Los resultados muestran que, durante el primer semestre de 2024, las temporadas iniciales de Grey's Anatomy registraron una mayor cantidad de vistas. La temporada 1 ocupa el primer lugar con 3.6 millones de vistas, seguida de la temporada 2 con 3.1 millones. A partir de ahí se observa, en términos generales, una disminución en la cantidad de vistas conforme aumenta el número de temporada, hasta llegar a la temporada 20 con 100 mil vistas.

Esto indica que durante el período analizado hubo una mayor concentración de visualizaciones en las temporadas más antiguas de la serie, mientras que las temporadas más recientes registraron una audiencia menor. Sin embargo, los datos disponibles permiten identificar este patrón de visualización, pero no determinar las razones que lo originaron.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [19]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [21]:
resultado = (
    tv_show[tv_show['title'] == "Grey's Anatomy"]
    .merge(season, left_on='id', right_on='tv_show_id',
           suffixes=('_show', '_season'))
    .merge(view_summary, left_on='id_season', right_on='season_id')
)

resultado = resultado[
    (resultado['duration'] == 'SEMI_ANNUALLY') &
    (resultado['start_date'].astype(str) == '2024-01-01') &
    (resultado['end_date'].astype(str) == '2024-06-30')
]

resultado = resultado[
    ['season_number', 'title_season', 'views']
].sort_values('views', ascending=False)

resultado

,season_number,title_season,views
0,1.0,Grey's Anatomy: Season 1,3600000
2,2.0,Grey's Anatomy: Season 2,3100000
4,3.0,Grey's Anatomy: Season 3,2900000
6,5.0,Grey's Anatomy: Season 5,2900000
8,4.0,Grey's Anatomy: Season 4,2900000
10,6.0,Grey's Anatomy: Season 6,2800000
12,7.0,Grey's Anatomy: Season 7,2700000
14,8.0,Grey's Anatomy: Season 8,2600000
16,9.0,Grey's Anatomy: Season 9,2500000
18,10.0,Grey's Anatomy: Season 10,2400000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [22]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [23]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [25]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = "/content/drive/MyDrive/Colab Notebooks"
os.chdir(DIR)

In [26]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [27]:
employees = db.collection('EMPLOYEE').stream()

data = []

for employee in employees:
    registro = employee.to_dict()
    registro['id'] = employee.id
    data.append(registro)

employee_df = pd.DataFrame(data).set_index('id')

employee_df

,emp_fname,emp_lname,emp_hiredate
id,,,
8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [28]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---